### PR TITLE
LUCENE-9268: Add some random tests to IndexWriter

### PR DESCRIPTION
**Pull request by dnhatn**
Add some tests that perform a set of operations randomly and concurrently on IndexWriter.

### CODE REVIEW

File: lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java [hunk 1]
```diff
@@ -3775,4 +3777,120 @@ public void testRefreshAndRollbackConcurrently() throws Exception {
...
+    }
+    for (Thread thread : threads) {
+      thread.join();
+    }
+    sm.maybeRefreshBlocking();
+    IOUtils.close(writer, sm, dir);
+  }
+
+  public void testRandomOperationsWithSoftDeletes() throws Exception {
+    Directory dir = newDirectory();
```
Comment (msokolov) [inline lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java]: can we use try-with-resources to ensure proper closure of Directory, IndexWriter, and SearcherManager on failure?

Comment (msokolov) [reply lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java]: Hmm I don't see where to officially +1 this in github, but thanks for sharing these tests!

**Comment by dnhatn** [comment]
@msokolov Thanks for looking. I pushed https://github.com/apache/lucene-solr/pull/1331/commits/9e9b53271372d6a884b8de019c4aec7fed7dd97f to use try-with-resources for them.

File: lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java [hunk 2]
```diff
@@ -3775,4 +3777,120 @@ public void testRefreshAndRollbackConcurrently() throws Exception {
...
+
+      @Override
+      public boolean keepFullyDeletedSegment(IOSupplier<CodecReader> readerIOSupplier) {
+        return keepFullyDeletedSegment;
+      }
+    });
+    try (Directory dir = newDirectory();
+         IndexWriter writer = new IndexWriter(dir, iwc);
+         SearcherManager sm = new SearcherManager(writer, new SearcherFactory())) {
+      Semaphore permits = new Semaphore(10 + random().nextInt(1000));
```
Comment (mikemccand) [inline lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java]: This is the number of operations (across all threads) we will execute?  Maybe rename it to `numOperations` or something?

File: lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java [hunk 3]
```diff
@@ -3775,4 +3777,120 @@ public void testRefreshAndRollbackConcurrently() throws Exception {
...
+    IndexWriterConfig iwc = newIndexWriterConfig();
+    AtomicInteger seqNo = new AtomicInteger(-1);
+    AtomicInteger retainingSeqNo = new AtomicInteger();
+    iwc.setSoftDeletesField("soft_deletes");
+    iwc.setMergePolicy(new SoftDeletesRetentionMergePolicy("soft_deletes",
+        () -> LongPoint.newRangeQuery("seq_no", retainingSeqNo.longValue(), Long.MAX_VALUE), newMergePolicy()));
+    try (Directory dir = newDirectory();
+         IndexWriter writer = new IndexWriter(dir, iwc);
+         SearcherManager sm = new SearcherManager(writer, new SearcherFactory())) {
+      Semaphore permits = new Semaphore(10 + random().nextInt(1000));
```
Comment (mikemccand) [inline lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java]: Also here?

File: lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java (hunk 2 above)
Comment (dnhatn) [reply lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java]: ++. Adjusted in https://github.com/apache/lucene-solr/pull/1331/commits/e3fd5ce0643d7a05fd0b530b20b58e7b5589cd2a.

**Comment by dnhatn** [comment]
@msokolov @mikemccand Thanks for review.

**Pull request closed without merging**